// @shopify/react-testing/build/cjs/TestWrapper.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

class TestWrapper extends React__default["default"].Component {
  constructor(...args) {
    super(...args);
    this.state = {};
    this.rootRef = null;

    this.setRef = ref => {
      this.rootRef = ref;
    };
  }

  // eslint-disable-next-line @shopify/react-prefer-private-members
  setProps(props) {
    this.setState({
      props
    });
  }

  render() {
    const {
      props
    } = this.state;
    const {
      children,
      render
    } = this.props;
    const rootElement = props ? /*#__PURE__*/React__default["default"].cloneElement(children, props) : children;
    return render( /*#__PURE__*/React__default["default"].createElement(TestInnerWrapper, {
      ref: this.setRef
    }, rootElement));
  }

}

class TestInnerWrapper extends React__default["default"].Component {
  render() {
    return this.props.children;
  }

}

exports.TestWrapper = TestWrapper;
